File: spring/src/main/java/com/example/demo/service/OrderService.java
package com.example.demo.service;

import com.example.demo.entity.order.Order;
import com.example.demo.entity.order.OrderProduct;
import com.example.demo.entity.product.Product;
import com.example.demo.entity.user.User;
import com.example.demo.repo.OrderRepository;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.stream.Collectors;

@Service
@Slf4j
public class OrderService {
    private final OrderRepository orderRepository;
    private final UserService userService;
    private final ProductService productService;

    public OrderService(OrderRepository orderRepository, UserService userService, ProductService productService) {
        this.orderRepository = orderRepository;
        this.userService = userService;
        this.productService = productService;
    }

    public Order findById(Long id) {
        return orderRepository.findById(id).orElse(null);
    }

    public List<Order> findAllOrderByLoginUser() {
        User user = userService.loginUserPrincipal().getUser();

        if (user == null) {
            log.warn("User info not found");
            return null;
        } else {
            return orderRepository.findAllOrderByOrderUserId(user.getId());
        }
    }

    @Transactional
    public Order orderProducts(List<Long> productIds) {
        User user = userService.loginUserPrincipal().getUser();

        if (user == null) {
            log.warn("User info not found");
            return null;
        } else {
            List<Product> products = productService.findAllByIds(productIds);
            List<OrderProduct> orderProducts = products
                    .stream()
                    .map(product -> OrderProduct.builder().product(product).build())
                    .collect(Collectors.toList());

            return orderRepository.save(
                    Order.builder()
                            .orderUser(user)
                            .orderList(orderProducts)
                            .build()
                    .prepareSave()
            );
        }
    }
}

File: spring/src/main/kotlin/com/example/demo/service/OrderService.kt
package com.example.demo.service

import com.example.demo.entity.order.Order
import com.example.demo.entity.order.OrderProduct
import com.example.demo.entity.user.User
import com.example.demo.repo.OrderRepository
import com.example.demo.security.JwtTokenProvider
import mu.KLogging
import org.springframework.security.core.context.SecurityContextHolder
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
class OrderService(
    private val orderRepository: OrderRepository,
    private val jwtTokenProvider: JwtTokenProvider,
    private val productService: ProductService
) {
    fun findById(id: Long): Order? {
        return orderRepository.findById(id).orElse(null)
    }

    fun findAllOrderByLoginUser(): List<Order> {
        return orderRepository.findAllOrderByOrderUserId(
            (SecurityContextHolder.getContext().authentication.principal as User).id!!
        )
    }

    @Transactional
    fun orderProducts(productIds: List<Long>): Order {
        val test = SecurityContextHolder.getContext().authentication.principal
        return orderRepository.save(
            Order(
                orderList = productService
                    .findAllByIds(productIds)
                    .map { product -> OrderProduct(product = product) },
                orderUser = (SecurityContextHolder.getContext().authentication.principal as User)
            ).prepareSave()
        )
    }

    companion object : KLogging()
}
